### ISSUE TITLE
Batch Jobs "complete" callback triggers after ONE job, not ALL

### ISSUE THREAD

**Issue by joemsak** [sidekiq/sidekiq#6110]
Am I completely misunderstanding the documentation? The docs say that "complete" and "success" will be called after "ALL" jobs complete (or succeed, respectively). But in my tests, the first job triggers complete:

```ruby
# spec/jobs/batch_jobs_spec.rb 

require "rails_helper"

class BatchJob
  attr_reader :items

  def initialize
    @items = (1..5).to_a
  end

  def start
    batch = Sidekiq::Batch.new
    batch.on(:complete, Callbacks, hello: :world)

    batch.jobs do
      items.each do |i|
        MyJob.perform_async(i)
      end
    end
  end

  class Callbacks
    def on_complete(status, options)
      puts "ALL JOBS COMPLETE!"
    end
  end

  class MyJob
    include Sidekiq::Worker

    def perform(i)
      puts i
      sleep 0.05
    end
  end
end

RSpec.describe BatchJob, :sidekiq_inline do
  it "test" do
    described_class.new.start
  end
end
```

OUTPUT:

```
Randomized with seed 52278
1
ALL JOBS COMPLETE!
2
3
4
5
.

Top 1 slowest examples (0.29171 seconds, 48.6% of total time):
  BatchJob test
    0.29171 seconds ./spec/jobs/batch_job_spec.rb:38

Finished in 0.60046 seconds (files took 2.67 seconds to load)
1 example, 0 failures

Randomized with seed 52278
```

Please advise on what I am doing wrong, misunderstanding in the docs, missing, etc. I have tried really hard to figure this out on my own before posting an issue. Thanks. 🙏 

Please see my answers below for the boilerplate stuff

---

Ruby version: 3.2.2
Rails version: 7.0.7.2

Sidekiq / Pro / Enterprise version(s):

```
gem "sidekiq", "~> 7.1.4"

source "https://enterprise.contribsys.com/" do
  gem "sidekiq-ent", "~> 7.1.2"
  gem "sidekiq-pro", "~> 7.1.4"`
end
```

Please include your initializer, sidekiq.yml, and any error message with the full backtrace.

```ruby
# config/initializers/sidekiq.rb
require "sidekiq"
require "sidekiq/web"
require "mail"

redis_config = {
  url: ENV.fetch("REDIS_SIDEKIQ_URL"),
  network_timeout: 5
}

unless Rails.env.test?
  Sidekiq::Enterprise.unique!
  Sidekiq::Client.reliable_push!
end

Sidekiq.configure_client do |config|
  config.redis = redis_config
end

Sidekiq.configure_server do |config|
  config.super_fetch!
  config.logger.level = Logger::WARN

  config.redis = redis_config

  config.periodic do |mgr|
    # see any crontab reference for the first argument
    # e.g. http://www.adminschoice.com/crontab-quick-reference
    mgr.register("0 4,9,11 * * *", "Admin::DailyAnalyticsWorker")
    mgr.register("0 0 * * *", "ExchangeRatesUpdatorWorker")
    mgr.register("*/15 * * * *", "ExpirationFinderWorker")
    mgr.register("*/15 * * * *", "UpcomingExpirationFinderWorker")
  end

  Mail.eager_autoload! if Rails.env.production? && defined?(Mail)
end

Sidekiq.strict_args!(false) unless ENV["SIDEKIQ_STRICT_ARGS"]
```

```YAML
# config/sidekiq.yml
:concurrency: <%= ENV.fetch("SIDEKIQ_CONCURRENCY", 5) %>
:queues:
  - mailers
  - default
  - calculators
```

**Comment by mperham** [comment]
I believe this bug was fixed in Pro 7.1.6. https://github.com/sidekiq/sidekiq/blob/main/Pro-Changes.md#716

**Comment by joemsak** [comment]
```
Randomized with seed 57660
1
2
3
4
5
ALL JOBS COMPLETE!
.

Top 1 slowest examples (0.30789 seconds, 49.6% of total time):
  BatchJob test
    0.30789 seconds ./spec/jobs/batch_job_spec.rb:38

Finished in 0.62088 seconds (files took 3.32 seconds to load)
1 example, 0 failures

Randomized with seed 57660
```
Yep, dang, sorry I missed being outdated, I felt like we had very recently ran an update so it didn't seem likely

**Comment by joemsak** [comment]
@mperham sorry, if I may ask, now that our batches appear to be working correctly:

Why the Web UI "Batches" page shows every single job in the batch as a "Batch" in this list? Shouldn't they be jobs inside the one batch that I created?

All the jobs listed as "batches"

![image](https://github.com/sidekiq/sidekiq/assets/98936/2b13ff71-3439-419c-bd92-0f62339cd422)

The one batch:

![image](https://github.com/sidekiq/sidekiq/assets/98936/665a4b04-9ced-42f1-92d8-12be481934c8)

Here is our code:

```ruby
  # ...
  def start_batches
    # ...
    batch = Sidekiq::Batch.new

    batch.description = "Foo"

    batch.jobs do
      MyJob.perform_bulk([[1, 2], [3, 4], etc...])
    end
  end
  # ...
```

**Comment by mperham** [comment]
I have no idea. I’ve never seen that behavior before. Maybe you fixed a bug in your code and those batches were created by the buggy code?

**Comment by joemsak** [comment]
@mperham I very very much appreciate your response and help with this issue, thank you. Well now that our batch is done, there is nothing listed anymore on the Batches page, and those were getting times of "Just now" in the "Started" column as I had "Live Polling" turned on, so I really think they were part of the one batch. 

Nothing in my code example looks out of whack to you? `perform_bulk` vs. `push_bulk` wouldn't make a diffrence, right?

**Comment by mperham** [comment]
Looks ok to me. I’m pretty sure perform_bulk calls push_bulk under the covers.

If you can give me a piece of code which reproduces the batch issue, I’ll be happy to debug and fix it.

**Comment by joemsak** [comment]
Okay @mperham thanks very much I'll write up an isolated reproduction. Maybe that will end up in me fixing my own problem 😅

**Comment by joemsak** [comment]
Hey again. Not asking for any aciton at this time, this is only a follow up for now, thanks!

I have not reproduced it in isolation but here are some screenshots of weird stuff going on in staging.

This is in the current default queue, all these batch callbacks for complete:
![image](https://github.com/sidekiq/sidekiq/assets/98936/4b2a13f9-5073-41f4-9458-52cb2e6688a5)

Here is the actual batch I purposely coded:
![Screenshot 2023-11-22 at 17-27-42 Sidekiq Enterprise](https://github.com/sidekiq/sidekiq/assets/98936/ee6d84d4-7cf8-4e60-8108-203c8c4dc9d9)

Here it is at the last page of the batches still remaining (the page count was much higher before the batch completed)
![Screenshot 2023-11-22 at 17-27-35 Sidekiq Enterprise](https://github.com/sidekiq/sidekiq/assets/98936/0df7a058-9765-481b-8ebc-053f0930c16d)

**Comment by joemsak** [comment]
@mperham I was just copying in the actual production code and I have a hunch, want to ask you:

If the code that executes inside the jobs has side-effects of triggering many more jobs, would those end up doing this, you think? I would not think so because I don't ever open another `batch.jobs` block, or reference the root branch `bid` as far as I know.

Here is the code and I'll point out where the side-effect jobs happen.

And here is the actual code:

```ruby
class PlanMigrationJob
  include Sidekiq::Worker

  def self.perform_in_batches(plan_migration_id)
    migration = PlanMigration.processing.find_by(id: plan_migration_id)

    return unless migration

    batch = Sidekiq::Batch.new

    batch.description = %(
      Plan Migration #{migration.id} from
      #{migration.from_plan_public_id} to
      #{migration.to_plan_public_id}
    ).squish

    batch.jobs do
      perform_bulk(migration.migration_map)
      # Example input: [[migration.affected_members[0].id, migation.id], [migration.affected_members[1].id, migration.id]]
    end
  end

  def perform(member_id, plan_migration_id)
    return unless valid_within_batch?

    migration = PlanMigration.processing.find_by(id: plan_migration_id)

    if migration.nil? 
    # Migration may have been canceled
    # (no longer in `.processing` scope)
    # at any time since its creation.
      Sidekiq::Batch.new(bid).invalidate_all
      return
    end

    limiter_id = "plan-migration-#{migration.id}"

    Sidekiq::Limiter.window(limiter_id, PlanMigration.batch_size, 1).within_limit do
      # Attempting to max Stripe ops at 50/s
      # estimated 6 ops/member, 50/6 == batch_size == 8
      member = migration.affected_members.find(member_id)
      purchaser = Action::PlanPurchaser.new(member:, plan: migration.to_plan)

      member.with_lock do

        # /!\ SIDE EFFECT JOBS HAPPEN HERE:
        if purchaser.purchase!


          migration.succeeded!(member:)
          # writes to DB in real time, no jobs
        else
          migration.failed!(member:, error_message: purchaser.error)
          # writes to DB in real time, no jobs
        end
      end
    end
  end
end
```

**Comment by mperham** [comment]
If you don't call `jobs` again, you shouldn't see any further jobs added to the batch. A batch job can create other jobs, those new jobs will not be associated with the batch.

**Comment by joemsak** [comment]
I found the culprit: we did in fact have a side-effect job that was being triggered, which was incorrectly coded about 6 years ago to use sidekiq batches to queue up a single job. LOL. 

So the 4,000+ jobs in the real batch were running code that coincidentally was calling up a new batch for a single job somewhere in the model callbacks. Removed that unnecessary batch and now our UI is tame. Thanks for your patience with me as I worked this issue out on my own. 

Hopefully at least if someone else out there has the same mystery batch problem as me, they find this useful for posterity.
